pkg/apis/keda/v1alpha1/triggerauthentication_types.go

```diff
@@ -83,6 +92,157 @@ type AuthEnvironment struct {
 	ContainerName string `json:"containerName"`
 }
 
+// Vault is used to authenticate using Hashicorp Vault
+// +k8s:openapi-gen=true
+type Vault struct {
+	Address        string              `json:"address"`
+	Authentication VaultAuthentication `json:"authentication"`
+
+	// +listType
+	Secrets []VaultSecret `json:"secrets"`
+
+	// +optional
+	Credential Credential `json:"credetial"`
+
+	// +optional
+	Role string `json:"role"`
+
+	// +optional
+	Mount string `json:"mount"`
+}
+
+// Credential defines the Hashicorp Vault credentials depending on the authentication method
+// +k8s:openapi-gen=true
+type Credential struct {
+	// +optional
+	Token string `json:"token"`
+
+	// +optional
+	ServiceAccount string `json:"serviceAccount"`
+}
+
+// VaultAuthentication contains the list of Hashicorp Vault authentication methods
+type VaultAuthentication string
+
+// Client authenticating to Vault
+const (
+	VaultAuthenticationToken      VaultAuthentication = "token"
+	VaultAuthenticationKubernetes                     = "kubernetes"
+	// VaultAuthenticationAWS                            = "aws"
+)
+
+// VaultSecret defines the mapping between the path of the secret in Vault to the parameter
+// +k8s:openapi-gen=true
+type VaultSecret struct {
+	Parameter string `json:"parameter"`
+	Path      string `json:"path"`
+	Key       string `json:"key"`
+}
+
 func init() {
 	SchemeBuilder.Register(&TriggerAuthentication{}, &TriggerAuthenticationList{})
 }
+
+// Authenticate returns a client connected to Vault
+func (v *Vault) Authenticate(logger logr.Logger) (*vaultApi.Client, error) {
+	config := vaultApi.DefaultConfig()
+	client, err := vaultApi.NewClient(config)
+
+	err = client.SetAddress(v.Address)
+	if err != nil {
+		return nil, err
+	}
+
+	token, err := v.token(client)
+	if err != nil {
+		return nil, err
+	}
+	client.SetToken(token)
+
+	lookup, err := client.Auth().Token().LookupSelf()
+	//If token is not valid so get out of here early
+	if err != nil {
+		return nil, err
+	}
+
+	renew := lookup.Data["renewable"].(bool)
+	if renew == true {
+		go v.renewToken(client, logger)
+	}
+
+	return client, nil
+}
+
+func (v *Vault) token(client *vaultApi.Client) (string, error) {
+	var token string
+
+	switch v.Authentication {
+	case VaultAuthenticationToken:
+		// Got token from VAULT_TOKEN env variable
+		if len(client.Token()) > 0 {
+			break
```

Comment:
this branch will return `""` for token, but then the only caller for this, `Authenticate()`, uses this value in `client.SetToken()`, won't that reset the token value?

Reply:
ohh thanks, missed it. 
https://github.com/nisan270390/keda/blob/a2aa58ce90439f0983f6ffb9e92a0ba3669f6a9d/pkg/handler/hashicorpvault_handler.go#L40
added a check if the token is not string empty